crashtracker/src/collector/crash_handler.rs

```diff
@@ -2,35 +2,268 @@
+    }
+}
+
+impl<const N: usize> Drop for SaGuard<N> {
+    fn drop(&mut self) {
+        // Restore the original signal actions
+        for &(signal, old_sigaction) in &self.old_sigactions {
+            unsafe {
+                let _ = signal::sigaction(signal, &old_sigaction);
+            }
+        }
+
+        // Restore the original signal mask
+        let _ = signal::sigprocmask(
+            signal::SigmaskHow::SIG_SETMASK,
+            Some(&self.old_sigmask),
+            None,
+        );
+    }
+}
+
+/// Opens a file for writing (in append mode) or opens /dev/null
+/// * If the filename is provided, it will try to open (creating if needed) the specified file.
+///   Failure to do so is an error.
+/// * If the filename is not provided, it will open /dev/null Some systems can fail to provide
+///   `/dev/null` (e.g., chroot jails), so this failure is also an error.
+/// * Using Stdio::null() is more direct, but it will cause a panic in environments where /dev/null
+///   is not available.
+fn open_file_or_quiet(filename: Option<&str>) -> anyhow::Result<RawFd> {
+    let file = filename.map_or_else(
+        || File::open("/dev/null").context("Failed to open /dev/null"),
+        |f| {
+            OpenOptions::new()
+                .append(true)
+                .create(true)
+                .open(f)
+                .with_context(|| format!("Failed to open or create file: {f}"))
+        },
+    )?;
+    Ok(file.into_raw_fd())
+}
+
+/// Non-blocking child reaper
+/// * If the child process has exited, return true
+/// * If the child process cannot be found, return false
+/// * If the child is still alive, or some other error occurs, return an error Either way, after
+///   this returns, you probably don't have to do anything else.
+fn reap_child_non_blocking(pid: Pid, timeout_ms: u64) -> anyhow::Result<bool> {
+    let timeout = Duration::from_millis(timeout_ms);
+    let start_time = Instant::now();
+
+    loop {
+        match waitpid(pid, Some(WaitPidFlag::WNOHANG)) {
+            Ok(WaitStatus::StillAlive) => {
+                if Instant::now().duration_since(start_time) > timeout {
+                    return Err(anyhow::anyhow!("Timeout waiting for child process to exit"));
+                }
+            }
+            Ok(_status) => {
+                return Ok(true);
+            }
+            Err(nix::Error::ECHILD) => {
+                // Non-availability of the specified process is weird, since we should have
+                // exclusive access to reaping its exit, but at the very least means there is
+                // nothing further for us to do.
+                return Ok(true);
+            }
+            _ => {
+                return Err(anyhow::anyhow!("Error waiting for child process to exit"));
+            }
+        }
+    }
+}
+
+/// Wrapper around the child process that will run the crash receiver
+/// TODO the execve arg coersion requires allocation, which we should avoid from a signal handler.
```

Comment:
I was going to comment on this, but you have identified it :+1: 
We are in the child process however this is still a signal handler :-/

Reply:
Yep, nothing wrong with preparing the args at signal handler install time instead, stuffing it into a global and freeing in atexit or such.

Reply:
Fixed -- these are now heap-allocated beforehand